store/store.go

```diff
@@ -385,18 +385,67 @@ func refreshMacaroon(user *auth.UserState) error {
 	return nil
 }
 
-// authenticateDevice will add the store expected Macaroon X-Device-Authorization header for device
-func (s *Store) authenticateDevice(r *http.Request) {
+// refreshUser will refresh user discharge macaroon and update state
+func (s *Store) refreshUser(user *auth.UserState) error {
+	err := refreshMacaroon(user)
+	if err != nil {
+		return err
+	}
+
 	if s.authContext != nil {
-		device, err := s.authContext.Device()
+		err = s.authContext.UpdateUser(user)
 		if err != nil {
-			logger.Debugf("cannot get device from state: %v", err)
-			return
-		}
-		if device.SessionMacaroon != "" {
-			r.Header.Set("X-Device-Authorization", fmt.Sprintf(`Macaroon root="%s"`, device.SessionMacaroon))
+			return err
 		}
 	}
+
+	return nil
+}
+
+// refreshDeviceSession will set or refresh the device session in the state
+func (s *Store) refreshDeviceSession() error {
+	if s.authContext == nil {
+		return fmt.Errorf("cannot get device from state: no authContext")
+	}
+
+	device, err := s.authContext.Device()
+	if err != nil {
+		return fmt.Errorf("cannot get device from state: %v", err)
+	}
+
+	serialAssertion, err := s.authContext.Serial()
+	if err != nil {
+		return err
+	}
+
+	nonce, err := RequestStoreDeviceNonce()
+	if err != nil {
+		return err
+	}
+
+	serialProof, err := s.authContext.SerialProof(nonce)
```

Comment:
We have the same issue here as we discussed with `key-proof`. We're covering this online.